MoltenVK/MoltenVK/GPUObjects/MVKDeviceMemory.mm

```diff
@@ -70,11 +68,11 @@
 	// Coherent memory does not require flushing by app, so we must flush now
 	// to support Metal textures that actually reside in non-coherent memory.
 	if (mvkIsAnyFlagEnabled(_vkMemPropFlags, VK_MEMORY_PROPERTY_HOST_COHERENT_BIT)) {
```

Comment:
What's the reasoning behind this? Wouldn't we want to map `MTLStorageModeShared` to `VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT|VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` and `MTLStorageModeManaged` to `VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT`? Later in the `flushToDevice` call we check that the storage is `MTLStorageModeManaged`, so wouldn't it be just better to check for that here and assume that any mappable memory is host visible (I believe validation would complain if it's not). Similar situation with `map` call.

Reply:
On macOS discrete GPU's, Metal does not always support `MTLStorageModeShared` for textures. Search the MoltenVK code base for references to `MTLStorageModeManaged` to see when we ignore `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` and force the use of `MTLStorageModeManaged`.

Metal has relaxed this restriction in later versions (see the `MVKPhysicalDeviceMetalFeatures::sharedLinearTextures` feature), so we might be able to rework the overall logic a bit, but that should already be taken into consideration wherever we force `MTLStorageModeManaged`.

Reply:
To get around the limitation of private-only `MTLHeaps` on non-Apple GPU's, I'm wondering if we could allocate a temporary managed `MTLBuffer` when a map call occurs, and do a BLIT from it to the aliased private `MTLBuffer` on flush? In this case, we would indicate that there is no shared memory options, so that the app would have to flush when it makes changes to the buffer content.

Reply:
I've been thinking about the issue for some time now. I'll dump my ideas here so I don't forget about them and we can decide.

First of, let's understand what Vulkan requires in terms of host visible coherent memory (which is our main issue due to non Apple silicon hardware). The spec states in https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#memory-device-properties the following:
`There must be at least one memory type with both the VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT and VK_MEMORY_PROPERTY_HOST_COHERENT_BIT bits set in its propertyFlags.`
and
`For any memory allocated with both the VK_MEMORY_PROPERTY_HOST_COHERENT_BIT and the VK_MEMORY_PROPERTY_DEVICE_COHERENT_BIT_AMD, host or device accesses also perform automatic memory domain transfer operations, such that writes are always automatically available and visible to both host and device memory domains.`

So we need to provide a host visible coherent type always and any memory writes either in host or device require no flushing to be visible on both.

The other restriction Vulkan imposes in https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#resources-association is the following:
`If buffer is a VkBuffer not created with the VK_BUFFER_CREATE_SPARSE_BINDING_BIT or VK_BUFFER_CREATE_PROTECTED_BIT bits set, or if image is a linear image that was not created with the VK_IMAGE_CREATE_PROTECTED_BIT bit set, then the memoryTypeBits member always contains at least one bit set corresponding to a VkMemoryType with a propertyFlags that has both the VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT bit and the VK_MEMORY_PROPERTY_HOST_COHERENT_BIT bit set. In other words, mappable coherent memory can always be attached to these objects.`

So, any buffer and linear texture (except if created with some flags) must have at least a memory type that is host visible and coherent. We can work around this by limiting what we expose through `vkGetPhysicalDeviceImageFormatProperties`, more on this later.

Now, looking at Metal texel buffers (MTLBuffers anyway) have a limit according to https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf which is 256MB, and is also limited by `MTLDevice.maxBufferLength`.

I do believe the way forward can be allocating private-only `MTLHeaps`, providing a fake memory on map, and flushing it to the `MTLHeap`. However, there are a few caveats that need addressing due to the previously mentioned points:
1. Host coherent memory requires that any write/read actually has the same value as the GPU counterpart memory. Why is this an issue? We cannot use map/unmap nor flush/invalidate Vulkan calls to do the flushing CPU memory to GPU memory and vice-versa. The user may map a VkDeviceMemory object, modify its contents by GPU commands, and read values from the mapped memory. I believe this case is not correctly handled in MoltenVK as of now either.
2. If we decided to use a `MTLBuffer` we are limited to either 256MB or `MTLDevice.maxBufferLength`. What happens if the mapped memory is bigger? How do we provide a contiguous memory for the user? Unsure how MoltenVK handles this at the moment.

Potential solutions to point 1 (will use GFXR as baseline since they've already addressed this issue and aligns to what I had in mind):
- https://github.com/LunarG/gfxreconstruct/blob/dev/USAGE_desktop_Vulkan.md#3-page_guard. In short, the idea is to trap the memory we provide to the user by changing the protection level and install a signal handler to catch accesses to that memory. Then we do our thing and flush/read to/from GPU in the handler.
- https://github.com/LunarG/gfxreconstruct/blob/dev/USAGE_desktop_Vulkan.md#4-userfaultfd. Unsure if this one is doable in macOS, I believe it should be, but need to put more time into it to correctly understand the viability.

Potential solutions to point 2 (when I use `the limitation` I'm refering to either 256MB or `MTLDevice.maxBufferLength` for this point, the smallest of those two):
- Expose a separate heap with the limitation as size that will be the one the user is allowed to allocate host visible coherent memory. The big downside to this approach is the potential low memory it would have.
- Limit linear texture sizes via `vkGetPhysicalDeviceImageFormatProperties` to not exceed the limitation. However, this does not limit users to allocate a bigger `VkDeviceMemory` than the limitation.
- Allocate host memory and provide that to the user. We also maintain a copy of this memory in one or multiple `MTLBuffer`s and do the flush/read to GPU using them. The big downside to this is the amount of memory we are using (3 times for the mapped region: one for the CPU, one for the buffers, one for the heap).

All of this is assuming we cannot create buffers with `MTLStorageShared` in non Apple silicon devices. I don't have a device with such set up, so I cannot verify this. I'm basing off MoltenVK's approach for those devices. If we had `MTLStorageShared` then a couple of things could be simplified, since we could just allocate buffers instead of heaps. But I believe at least point 2 would require addressing.

Any thoughts, comments or concerns are welcome!

Reply:
Most of this is managing syncing non-Apple GPU memory. I think there will always be situations where someone opens a memory map, leaves it open, and constantly makes changes to it, so we may always have to accept that we will have incomplete solutions.

> `There must be at least one memory type with both the VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT and VK_MEMORY_PROPERTY_HOST_COHERENT_BIT bits set in its propertyFlags.`

It might be possible to modify this within the spec with something like "**_If the `VK_KHR_portability_subset` extension is not enabled, ..._**", allowing us to back-out of certain standard Vulkan requirements, and then modify CTS to not fail when this happens.